tests/dismod/db/test_wrapper.py

```diff
@@ -0,0 +1,81 @@
+import numpy as np
+import pandas as pd
+
+import pytest
+
+from sqlalchemy.exc import StatementError
+
+from cascade.dismod.db.wrapper import DismodFile, _get_engine
+
+
+@pytest.fixture
+def engine():
+    return _get_engine(None)
+
+
+@pytest.fixture
+def base_file(engine):
+    dm_file = DismodFile(engine, {"howdy": float}, {"there": int})
+    dm_file.make_densities()
+    ages = pd.DataFrame({"age": np.array([6.0, 22.0, 48.0])})
+    dm_file.age = ages
+    dm_file.time = pd.DataFrame({"time": [1997.0, 2005.0, 2017.0]})
+    dm_file.integrand = pd.DataFrame({"integrand_name": ["prevalence"]})
+
+    return dm_file
+
+
+def test_wrong_type(base_file):
+    ages = pd.DataFrame({"age": np.array(["strings", "for", "ages"])})
+    base_file.age = ages
+    with pytest.raises(StatementError):
+        base_file.flush()
```

Comment:
We could do explicit validation elsewhere, but it's tricky. Pandas dataframes don't (as far as I know) provide a way to hook behavior onto column mutations. That means that I can't really know what the contents of a given DataFrame are until I actually check which I only do at flush time. Like the dirty bit stuff above, you could solve this by writing a complex proxy. Or changing the API such that you had to explicitly set a whole DataFrame at a time so there was a place to do these checks. Something like this:

data_frame = dbfile.get_data("key")
dbfile.set_data("key", data_frame)

That makes this validation stuff much easier, but it's a less pleasant API. I think I like the current version, even though it throws late rather than early in this case.

Reply:
Let's not mess with the Pandas api but focus on throwing a beautiful error on flush.